linkerd/proxy/detect/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+use futures::{try_ready, Future, Poll};
+use linkerd2_error::Error;
+use linkerd2_io::{BoxedIo, Peek};
+use linkerd2_proxy_core as core;
+
+/// A strategy for detectin values out of a client transport.
+pub trait Detect<T> {
+    type Target;
+
+    /// If the target can be determined by the target alone (i.e. because it's
+    /// known to be a server-speaks-first target), Otherwise, the target is
+    /// returned as an error.
+    fn detect_before_peek(&self, target: T) -> Result<Self::Target, T>;
```

Comment:
Hmm, I'm a little bit on the fence about using `Result` to indicate incomplete rather than erroneous states, but the `Conditional` type has kind of scared me out of defining new enums for cases that can reasonably use `Result`s...this is probably the best approach...

Reply:
And it's not _wrong_, here, really. It's like how sender uses Result to return the item if the receiver is gone.

Reply:
yeah, i think this is probably the best thing here, i just had to think about it for a minute.